### PR TITLE
Rearrange data objects

### PR DESCRIPTION
**Pull request by matthewturk**
This is a strict refactoring which moves objects from selection data objects into a subdirectory, `selection_objects`.  It is in support of #2827 .

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Looks like a continuation of https://github.com/yt-project/yt/pull/2823, right ?

File: yt/frontends/adaptahop/data_structures.py [hunk 1]
```diff
@@ -23,6 +22,7 @@
 from yt.units import Mpc
 from yt.utilities.cython_fortran_utils import FortranFile
 
+from ...data_objects.selection_objects.base_objects import YTSelectionContainer
```
Comment (neutrinoceros) [inline yt/frontends/adaptahop/data_structures.py:25]: wait this is valid python??

**Comment by matthewturk** [comment]
Yup!

File: yt/frontends/ytdata/data_structures.py [hunk 2]
```diff
@@ -625,7 +624,7 @@ def get_data(self, fields=None):
         if len(fields_to_get) == 0 and len(fields_to_generate) == 0:
             return
         elif self._locked:
-            raise GenerationInProgress(fields)
+            raise YTGenerationInProgress(fields)
```
Comment (neutrinoceros) [inline yt/frontends/ytdata/data_structures.py:627]: Couldn't figure out where ` GenerationInProgress` was supposed to be defined, so isn't this line a bugfix ?

File: yt/utilities/flagging_methods.py [hunk 3]
```diff
@@ -8,13 +8,16 @@
 class RegisteredFlaggingMethod(type):
```
Comment (neutrinoceros) [inline yt/utilities/flagging_methods.py:8]: Is there a reason this class isn't removed in https://github.com/yt-project/yt/pull/2823 ?

File: yt/utilities/object_registries.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# These are some of the data object registries that are used in different places in the code.  Not all of the self-registering objects are included in these.
```
Comment (neutrinoceros) [inline yt/utilities/object_registries.py:1]: this line violates flake8 E501 (line too long) that I'm trying to reintroduce in #2796
```suggestion
# These are some of the data object registries that are used in different places in the code. 
# Not all of the self-registering objects are included in these.
```

File: yt/utilities/flagging_methods.py (hunk 3 above)
Comment (matthewturk) [reply yt/utilities/flagging_methods.py:8]: Yes -- it's super duper localized, and a candidate for removal at a later time.

File: yt/frontends/ytdata/data_structures.py (hunk 2 above)
Comment (matthewturk) [reply yt/frontends/ytdata/data_structures.py:627]: Well, sorta.  I actually renamed *and* moved it.  It used to just be `GenerationInProgress` but now I moved it to `yt/utilities/exceptions.py` and renamed it with the prefix.

File: yt/utilities/object_registries.py (hunk 4 above)
Comment (matthewturk) [reply yt/utilities/object_registries.py:1]: Oops, this was my mistake; I actually word-wrapped it and then messed up with my changeset.

File: yt/frontends/adaptahop/data_structures.py (hunk 1 above)
Comment (matthewturk) [reply yt/frontends/adaptahop/data_structures.py:25]: Yeah, but I actually am going to change these; I would rather absolute imports than relative for these.  I missed that pycharm did this.  (Yeah, that's right, I used a tool to help me refactor!)

File: yt/frontends/ytdata/data_structures.py (hunk 2 above)
Comment (neutrinoceros) [reply yt/frontends/ytdata/data_structures.py:627]: Any chance we'd except some downstream code to rely on catching this ? If so, maybe we should apply a similar pattern as here https://github.com/yt-project/yt/pull/2818

**Comment by matthewturk** [comment]
I'll be honest: it is possible!  And, more to the point, it's not *that*
big of a deal to me to rename it back, so I'll go ahead and do that.

File: yt/frontends/adaptahop/data_structures.py (hunk 1 above)
Comment (neutrinoceros) [reply yt/frontends/adaptahop/data_structures.py:25]: >  (Yeah, that's right, I used a tool to help me refactor!)
*screams in vanilla emacs*

Comment (matthewturk) [reply yt/frontends/adaptahop/data_structures.py:25]: Well I *tried* to get `rope` to work but couldn't!

**Comment by neutrinoceros** [comment]
Aw, too bad, I thought this change was very welcome. After wandering the codebase for a while, one gets accustomed to having custom exceptions called `YTStuff`, so this one really looks like it's coming from elsewhere when it isn't.

**Comment by matthewturk** [comment]
I'll issue another PR that changes the name.

**Comment by neutrinoceros** [comment]
Weird that the slash command didn't satisfy Travis. I bet it's a version clash between the two.
I added version printing for travis here #2833 (PR for slash command repo linked therein)
Note that I don't know what version of isort is run by the slash command but it looks like it's _older_ than 5.2.1 (the current pinned version in the main repo). There are already more recent versions but they yield consistent results with 5.2.1 for now.

**Comment by matthewturk** [comment]
Looks like flynt isn't the right version or something.  Manual time!

**Comment by matthewturk** [comment]
I had to resolve a conflict, which I *think* was ... not a big deal ... but if somebody with some better git facility could take a look I'd appreciate it.

**Comment by neutrinoceros** [comment]
I'll give it a look !

File: yt/data_objects/selection_objects/base_objects.py [hunk 5]
```diff
@@ -0,0 +1,1431 @@
...
+                        # If the field returns an ndarray, coerce to a
+                        # dimensionless YTArray and verify that field is
+                        # supposed to be unitless
+                        fd = self.ds.arr(fd, "")
+                        if fi.units != "":
+                            raise YTFieldUnitError(fi, fd.units)
+                    except UnitConversionError:
+                        raise YTFieldUnitError(fi, fd.units)
+                    except UnitParseError:
+                        raise YTFieldUnitParseError(fi)
```
Comment (neutrinoceros) [inline yt/data_objects/selection_objects/base_objects.py:271]: lost in between mergers
```suggestion
                    except UnitConversionError as e:
                        raise YTFieldUnitError(fi, fd.units) from e
                    except UnitParseError as e:
                        raise YTFieldUnitParseError(fi) from e
```

**Review by neutrinoceros — APPROVED** [review]
alright that's everything wrong I found with the latest merge.
Btw, `base_objects.py` is only has `YTDataSelectionContainer<DIM>` classes in it right ?
if so, maybe the file should be named `data_selection_containers.py` instead ?

**Review by cphyc — APPROVED** [review]

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
Dang it, conflicts again after sending #2610 in, sorry @matt

**Comment by matthewturk** [comment]
fixed conflicts

https://www.youtube.com/watch?v=r51EomcIqA0

**Comment by neutrinoceros** [comment]
Exceedingly efficient vid !

**Pull request merged**